### ISSUE TITLE
MDSplus.Range() issue: depending on the delta step, the last element of the original array is included in the final output.

### ISSUE THREAD

**Issue by santorofer** [MDSplus/mdsplus#1802]
In MDSplus Range() method, depending on the value of the delta step to divide the array, the last element of the array is included in the array output by Range(), which it's not the correct answer for Range().

See the following examples:

```python
>>> Range(0., 9., 1.4999).data()
array([ 0.    ,  1.4999,  2.9998,  4.4997,  5.9996,  7.4995, 8.9994])
>>> Range(0., 9., 1.5001).data()
array([ 0.    ,  1.5001,  3.0002,  4.5003,  6.0004,  7.5005, 9.0006])
```

In other words, if the remainder of the division between the last element of the array and the delta step is less than 0.5, then the array contains the correct elements:
```python
>>> Range(0., 9., 1.1).data()
array([0. , 1.1, 2.2, 3.3, 4.4, 5.5, 6.6, 7.7, 8.8])
```
where the remainder of the division is less that 0.5:
```python
>>> 9/1.1
8.181818181818182
```
On the other hand, if the remainder is 0.5 or greater:
```python
>>> Range(0., 9., 1.2).data()
array([0. , 1.2, 2.4, 3.6, 4.8, 6. , 7.2, 8.4, 9.6])
>>> 9/1.2
7.5

>>> Range(0., 9., 1.3).data()
array([0. , 1.3, 2.6, 3.9, 5.2, 6.5, 7.8, 9.1])
>>> 9/1.3
6.9230769230769225
```

**Comment by zack-vii** [comment]
Just to make sure I understand you wish for range to act exactly like e.g. python's range. 

> which it's not the correct answer for Range().

What definition are you referring to. It seems - based on the source code - the current behavior is intentional. It is caused by TdiNint in tdishr/TdiDtypeRange.c line 174:
```c
  if (STATUS_OK && nelem.pointer->dtype != DTYPE_L) {
    status = TdiNint(&nelem, &nelem MDS_END_ARG);
    if (STATUS_OK && nelem.pointer->dtype != DTYPE_L)
      status = TdiLong(&nelem, &nelem MDS_END_ARG);
  }
```

**Comment by tfredian** [comment]
One can debate whether the behavior of some of the tdi intrinsics is "correct" but remember that these functions were implemented decades ago even before languages such as python were created. It is generally not a good idea to change the behavior of tdi unless there is an obvious bug. Users learned the behavior of these functions and designed their expressions to compute their physical measurements based on the existing behavior and these expressions are often stored with the data.

**Comment by joshStillerman** [comment]
Tom -
  I asked Fernando to put in this issue to open the discussion.  The real issue is:  Range sometimes returns a value > its end value. The example
```
>>> Range(0., 9., 1.4999).data()
array([ 0.    ,  1.4999,  2.9998,  4.4997,  5.9996,  7.4995, 8.9994])
>>> Range(0., 9., 1.5001).data()
array([ 0.    ,  1.5001,  3.0002,  4.5003,  6.0004,  7.5005, 9.0006])
```
demonstrates this.  I agree we should be concerned about changing its behavior, but I think the 2nd answer above really needs to be:
```
>>> Range(0., 9., 1.5001).data()
array([ 0.    ,  1.5001,  3.0002,  4.5003,  6.0004,  7.5005])
```

**Comment by joshStillerman** [comment]
A conversation ending in a solution for the problem.
```
Josh 9:36 AM
I know that

I think range should return values <= to its end value
0 : 3 : 1 --> 4 values
0 : 3 : 1.1 --> 3 values
Tom Fredian 9:39 AM
well if you want to go through all the existing cmod data and determine that such a change wouldn't break anything...
as Timo mentioned the code explicitly behaves that way intentionally.
Josh 9:39 AM
do you agree that returning : 0, 1.1, 2.2, 3.3 is correct
incorrect
as 3.3 is clearly not in the range 0 : 3
Tom Fredian 9:41 AM
there are lots of tdi functions that behave somewhat different than one might expect but I think users worked around them if they needed to.
Ask Ken what he thinks.
Josh 9:42 AM
we can put logic in the device driver in question to remove the extra point, which only shows iup with particular values of dt and segment length/
However this seems kind of broken.
Tom Fredian 9:43 AM
or perhaps we add a prange() fun.
or pyrange()
Josh 9:43 AM
we could make a new Range1  object with different (more correct) behavior and use that, but that would be painful
pyrange()  is different
I am not suggesting 0 : 3 : 1  , only return 0, 1, 2
as it would in python
Tom Fredian 9:44 AM
true
Josh 9:44 AM
I am suggesting 0 : 3 : 1 return 0, 1, 2, 3
but
0 : 3 :1.1  return 0, 1.1, 2.2
Tom Fredian 9:45 AM
cliprange()
Josh 9:45 AM
I like that better than
Tom Fredian 9:45 AM
probably easy to do in python.
Josh 9:45 AM
1 - changing range
or
2 - making the code awkward
3 - bon idea to write it in python
or we could just write clip(array, min, max)
and use it in the device
Tom Fredian 9:46 AM
yep
Josh 9:46 AM
clip(range(start, end, dt), start, end)
Tom Fredian 9:47 AM
sounds good to me
Josh 9:48 AM
if clip is a TDI/Python function, then do we need a Clip python object analogous to Range ?
Tom Fredian 9:49 AM
don't think so.
Josh 9:50 AM
so right now we write Signal(y,None, Range(...))
what is the syntax for referring to an expression in that context?
Tom Fredian 9:51 AM
I would think Signal(y,None,clip(Range(...),min,max)) would work. (edited) 
Josh 9:51 AM
python won't know what to do with that
could be
Signal(y, None, Function('clip', Range(...))  - not sure Function is the correct object name
Tom Fredian 9:56 AM
hmmm, I forget how to do that. You'd need to do something similar for Signal($value*a+b,array,dim)
Josh 9:56 AM
OK we will work it out or make Clip object
that sticks in the function
Tom Fredian 9:57 AM
making a Clip object would be complicated I think.
That would require changes to the tdishr code, header files, oo code (python, java, c++) etc.
Josh 9:58 AM
could be, we must have a function object of some sort, so we can use that. or make something that wraps that
Don't think so
it could be completely pythonic
oh you are correct
the trivial way I was thinking of, would not decompile
Tom Fredian 9:59 AM
I can play with it a bit.
Josh 9:59 AM
I will paste this conversation into the issue and assign it to Fernando.  He can start with the python code that impliments Clip  for tdi```

**Comment by tfredian** [comment]
To utilize this clip function in the python device code you can use something similar to the following:

sig=Signal(y,None,Data.compile('clip($)',Range(start,end,inc)))

**Comment by zack-vii** [comment]
I think if we decide to add clip we should probably make it a built in. there might even be a built in already. maybe CULL can do that. 
```tdi
DATA(0.:3.:1.1)
[0.,1.1,2.2,3.3]
CULL([0.,3.],,0.:3.:1.1)
[0.,1.1,2.2]
```

**Comment by zack-vii** [comment]
everyone agrees?

**Comment by tfredian** [comment]
Great idea!!!
Sounds good to me... To use cull in a python device you could simply do the following:
```
>>> r=Range(0.,9.,1.2) 
>>> sig=Signal(42,None,CULL(Array([r.begin,r.ending]),None,r))
```
which produces:
```
>>> sig.decompile()
'Build_Signal(42, *, CULL([0D0,9D0], *, 0D0 : 9D0 : 1.2D0))'
>>> sig.dim_of().data()
array([ 0. ,  1.2,  2.4,  3.6,  4.8,  6. ,  7.2,  8.4])

**Comment by santorofer** [comment]
Ok, I think I found something else. I was implementing CULL() when I noticed the following: if I reduce the values of Tom's example 10 times smaller, in other words, for a range between [0.,1.], i.e for a Range(0., 0.9, 0.12), then,
```python
>>> Signal(1,None,CULL(Range(0.,.9,None),None,Range(0.,.9,0.12))).dim_of().data()
array([0.  , 0.12, 0.24, 0.36, 0.48, 0.6 , 0.72, 0.84, 0.96])
>>> Signal(1,None,CULL(Range(0.,.9,None),None,Range(0.,.9,0.13))).dim_of().data()
array([0.  , 0.13, 0.26, 0.39, 0.52, 0.65, 0.78, 0.91])
>>> Signal(1,None,CULL(Range(0.,.9,None),None,Range(0.,.9,0.14))).dim_of().data()
array([0.  , 0.14, 0.28, 0.42, 0.56, 0.7 , 0.84])
```

The same, for example, if I just get the dimension from CULL() itself:

``` python
>>> CULL(Range(0.,.9,None),None,Range(0.,.9,0.13)).getArgumentAt(2).data()
array([0.  , 0.13, 0.26, 0.39, 0.52, 0.65, 0.78, 0.91])
```

Strange, isn't it? Unless I'm missing something, that that could be also the case.

So, I implemented the python device code simply by doing the clipping that we were discussing at the beginning.

**Comment by joshStillerman** [comment]
Do you get the same result if the 1st argument to cull is Array(min, max) ?On Sep 25, 2019 7:32 PM, Fernando Santoro <notifications@github.com> wrote:Ok, I think I found something else. I was implementing CULL() when I noticed the following: if I reduce the values of Tom's example 10 times smaller, i.e for a Range(0., 0.9, 0.12), then,
>>> Signal(1,None,CULL(Range(0.,.9,None),None,Range(0.,.9,0.12))).dim_of().data()
array([0.  , 0.12, 0.24, 0.36, 0.48, 0.6 , 0.72, 0.84, 0.96])
>>> Signal(1,None,CULL(Range(0.,.9,None),None,Range(0.,.9,0.13))).dim_of().data()
array([0.  , 0.13, 0.26, 0.39, 0.52, 0.65, 0.78, 0.91])
>>> Signal(1,None,CULL(Range(0.,.9,None),None,Range(0.,.9,0.14))).dim_of().data()
array([0.  , 0.14, 0.28, 0.42, 0.56, 0.7 , 0.84])
The same, for example, if I just get the dimension from CULL() itself:
>>> CULL(Range(0.,.9,None),None,Range(0.,.9,0.13)).getArgumentAt(2).data()
array([0.  , 0.13, 0.26, 0.39, 0.52, 0.65, 0.78, 0.91])
Strange, isn't it? Unless I'm missing something, that that could be also the case.
So, I implemented the python device code simply by doing the clipping that we were discussing at the beginning.

**Comment by santorofer** [comment]
That’s a good point, I didn’t use the array directly in the first argument as you mention. I will do that tomorrow.

**Comment by tfredian** [comment]
Yes. 

Signal(1,None,CULL(Array([0.,.9]),None,Range(0.,.9,0.12))).dim_of().data()                
array([ 0.  ,  0.12,  0.24,  0.36,  0.48,  0.6 ,  0.72,  0.84])

The problem with using the Range is that the default increment for Range is 1:

Range(0.,.9).data()
array([0.,1.])

**Comment by santorofer** [comment]
Yes, I noticed the [0.,1] range that you mention.
Ok, it's working fine now, I just modified the python device code.
There is only one detailed, in that when the remainder of segment_1_lenght/decimator is zero, then the last element of the result from CULL is the original last element of the range, which will then appear as duplicate because it will also be part of the first element of the second segment, and so on. So, I just put a logic in the code to remove that last element, so that it appears only one in each segment. By doing that I'm able to see good behavior for any decimation values.
